### DISCUSSION TITLE
Additional mods on Android

### DISCUSSION THREAD

**Discussion by stuffi3000** [exult/exult#522] [General]
Hello everyone,

I have recently started playing Ultima 7 on Android using Exult and the mods which can be enabled in the launcher. However, I would really like the install an alternative font (http://exult.info/forum/viewtopic.php?p=1711205#p1711205) as the current one is really bothering my eyes on the small Android screen.

Based on some searchers in the discussions here I would that the app currently saves all its data in, well, the Android data folder. Since I'm on Android 14, on a non-rooted phone, I can't even access that path using Shizuku (local Android ADB API) + ZArchiver. I therefore conclude that I can't manually install mods on Android? 

If that is indeed the case, I would like to suggest to either move the Android data into /storage/emulated/0/Android/data to allow easier editing, or add the possibility for manual installation of mods in the launcher. That would make manualling installing mods on Android possible/easier and would also help people to transfer their save files without having to make a backup through ADB.

Thank you very much!

**Comment by DominusExult** [comment]
Someone more versed in Android would need to look into doing this.
The one who made the port ran out of time to do more and I can barely maintain it ;(

**Reply by stuffi3000** [reply]
Oh, I didn't know that...
Well, it's already an awesome thing to have a maintained Android port to begin with, thank you very much for that :)

**Comment by wench** [comment]
This has me thinking that we possibly should have an extra mods directory for all platforms in a user writable location. As it would be for all platforms we'd just need to #ifdef the android version to use the appropriate directory. Nothing too complex.

**Comment by wench** [comment]
I should add extra clarification i would add a system wide shared mods path that is not normally user writable, and an additional user writeable path the user can download mods to if they do not have access to the shared path (because of no permissions or exult is in an immutable package and could came with mods).

**Reply by i30817** [reply]
There is a workaround to get mods running in android because the current launcher allows zips for mods into the internal APK directory. It's supposed to be used for keyring and sifixes and the launcher checks those directories and the cfg file exist...

But it doesn't forbid some more directories and cfg files in the same zip.

This is how you can get the u6 mod running in android. Beware you have to make sure the u6 mod files and the u7 files are lower cased and move the cfg file a single dir up like the others.

Then you zip everything (In this case, since the mod is for blackgate forge of virtue (blackgate only doesn't work), keyring and u6 dirs, and cfgs) and make the launcher 'install' the blackgate fov mod with that zip. If everything went correctly, the u6 mod should be available in the mod menu  and in game the houses not have missing walls (that's a overwrite problem from case conflicts).

**Comment by DominusExult** [comment]
hmm, I'll have to make sure to lowercase all files in the copy of the U6 mod on our download page.

**Reply by i30817** [reply]
Eh. I think it's not so much lower cased, but 'same cased' as the original game files (to replace them). You can do that, for neatness, but the real solution is the PR for the new file IO, because not everyone will have the original game files lowercased. Unless you make the launcher android App ensure that ofc.

**Comment by BaldrIsCoding** [comment]
Did you find a solution? I made a similar post like an idiot without checking this one before doing that. I'd like to be able to change the fonts as well since I'm not a root user too.

**Reply by i30817** [reply]
The solution is to include multiple mods whose case matches the case of the original of files in the zip the launcher allows you to install as a mod. For instance, I lowercase with a Linux command (termux will do it, search for the command in this project bug report page) all the files in the data files for each game, then I download and unzip the zip for (for instance) keyring mod, lowercase that too, download the zip for the u6 mod (a blackgate+forge of virtue mod iirc), then unzip that too and lower case it. Then I rezip the data files, and install them with the launcher, then I move the data files for the u6 mod to the directory holding the keyring mod (the u6 mod has the additional requirement to move the cfg file to the parent dir because the mod creator screwed that up), then zip up that directory and use the zip to install the keyring mod in the launcher.

Since the launcher only checks if the zip contains the keyring mod, not forbid any other mod, you can use it as a trojan horse to install other mods. Just have to be careful that the game selected works with the mods (for example, u6 mod will not work with the simple blackgate game, just blackgate + fov).
